File: main/src/ui/java/freedom/app/parser/MessageParser.kt
package freedom.app.parser

object MessageParser {

    /**
     * Detects the message type from the raw string and delegates to the appropriate parser.
     * Returns null if the message is empty or unrecognisable.
     */
    fun parse(raw: String, senderAddress: String? = null): ParsedMessage? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null

        return when {
            trimmed == "PING"             -> ParsedMessage(MessageType.PING,  "", senderAddress)
            trimmed == "PONG"             -> ParsedMessage(MessageType.PONG,  "", senderAddress)
            trimmed.startsWith("SRCH:")   -> SearchMessageParser.parse(trimmed, senderAddress)
            trimmed.startsWith("INFRA:")  -> InfraMessageParser.parse(trimmed, senderAddress)
            else                          -> TextMessageParser.parse(trimmed, senderAddress)
        }
    }
}

File: ios/Freedom/Core/Parser/MessageParser.swift
// ═════════════════════════════════════════════════════════════════════════════
//  MessageParser — Router dispatcher for message parsing
// ═════════════════════════════════════════════════════════════════════════════

import Foundation

enum MessageParser {

    /// Detects the message type from the raw string and delegates to the appropriate parser.
    static func parse(_ raw: String, senderAddress: String? = nil) -> ParsedMessage? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        if trimmed == "PING" {
            return ParsedMessage(type: .PING, content: "", senderAddress: senderAddress)
        }
        if trimmed == "PONG" {
            return ParsedMessage(type: .PONG, content: "", senderAddress: senderAddress)
        }
        if trimmed.hasPrefix("SRCH:") {
            return SearchMessageParser.parse(trimmed, senderAddress: senderAddress)
        }
        if trimmed.hasPrefix("INFRA:") {
            return InfraMessageParser.parse(trimmed, senderAddress: senderAddress)
        }
        return TextMessageParser.parse(trimmed, senderAddress: senderAddress)
    }
}
